File: leetcode/2369.检查数组是否存在有效划分.go
func validPartition(nums []int) bool {
    n := len(nums)
    f := make([]int, n + 1)
    f[0] = 1
    for i := 0; i < n; i ++ {
        if i >= 1 && nums[i] == nums[i - 1] && f[i - 1] == 1{
            f[i + 1] = 1
            continue
        }
        if i >= 2 && nums[i] == nums[i - 1] && nums[i - 1] == nums[i - 2] && f[i - 2] == 1 {
            f[i + 1] = 1
            continue
        }
        if i >= 2 && nums[i] == nums[i - 1] + 1 && nums[i - 1] == nums[i - 2] + 1 && f[i - 2] == 1{
            f[i + 1] = 1
            continue
        }
    }
    return f[n] == 1
}

File: leetcode/2369.检查数组是否存在有效划分.py
class Solution:
    def validPartition(self, nums: List[int]) -> bool:
        n=len(nums)
        dp=[False]*(n+1)
        dp[0]=True
        for i in range(1,n):
            if dp[i-1] and nums[i]==nums[i-1]:
                dp[i+1]=True
            if i>1 and dp[i-2] and nums[i]==nums[i-1]==nums[i-2]:
                dp[i+1]=True
            if i>1 and dp[i-2] and nums[i]==nums[i-1]+1 and nums[i-1]==nums[i-2]+1:
                dp[i+1]=True
        return dp[n]